### PR TITLE
`get_book_times` helper function

### PR DESCRIPTION
**Pull request by jlashner**
This PR adds a new function to the imprinter that takes in an obs book (assuming a timing system), and returns a full gap-less list of timestamps that should be contained in the book.

This only takes 15 seconds to run for a 2 hour run for a two-hour timestream and I think could both:
-  simplify the gap-filling logic of the bookbinder, and allow for frame pre-allocation which may speed up current run-times
- Help the bookbinder determine whether there are missing samples at the start/end of files and frames

Relevant: #353

### CODE REVIEW

File: sotodlib/io/imprinter.py
Comment (mhasself) [inline sotodlib/io/imprinter.py:692]: So this does the "trimming" for you ... maybe mention that.

This is an opportunity to report how much time has been trimmed from the various filesets to make them all the same length... but I imagine that's been checked carefully elsewhere.

File: sotodlib/io/imprinter.py
Comment (algadoreo) [inline sotodlib/io/imprinter.py:703]: A downside to using only one wafer to determine the list of timestamps is that the filled-in samples may not line up exactly with the samples that exist in other wafers. It probably doesn’t matter since we’ll flag those samples anyway, but I thought I should point that out.

I realize that checking all the wafers adds computational time, so maybe it could exist as a user-option to perform a more “comprehensive” search: the times for each wafer is assembled and trimmed, then look for unique times in the whole set. (Uniqueness only works when the timing system is on, but the correct fallback for no timing system is to make single-wafer books where it’s okay if samples don’t line up.) The vector of unique times is then filled with the missing times.

File: sotodlib/io/imprinter.py
Comment (algadoreo) [inline sotodlib/io/imprinter.py:764]: Would median be better here? I think a large gap (or a large number of smaller gaps) might skew the mean too much upward. On the other hand, the median would be affected by a large variance, but a timing system should keep that in check, right?

Also, probably an unlikely scenario, but if for some reason you get two samples close to each other (dts/dt < 0.5), then I think you end up with -1 missing samples, which would break the subsequent code. Maybe add a max(0, missing) or something to be safe?

File: sotodlib/io/imprinter.py
Comment (algadoreo) [inline sotodlib/io/imprinter.py:747]: Perhaps `fill_time_gaps` to make it clear it’s for time gaps?

**Review by algadoreo — COMMENTED** [review]
Thanks for this! It’ll make my life a lot easier once this is merged.

File: sotodlib/io/imprinter.py
Comment (jlashner) [reply sotodlib/io/imprinter.py:692]: Right. where do you think would be a good place to report this? should it make its way into the M_index file?

File: sotodlib/io/imprinter.py
Comment (jlashner) [reply sotodlib/io/imprinter.py:703]: Not sure I understand you here. Are you talking about if the slots are not actually co-sampled from a timing system?

It kind of seems like you are talking about a case where the filled timestamps are something like:
`[0, 1, 2, 3, 4, 5]`, but one sample happens to have an extra timestamp like `[0, 1, 2, 3, 3.5, 4, 5]`.
I think the correct action here would be to not include the `3.5` sample of the second array, which I think this method will actually correctly handle.

I'm probably misinterpreting you're concern though... do you have like an example of timestamp arrays that demonstrates your thought?

Comment (algadoreo) [reply sotodlib/io/imprinter.py:703]: I mean something like this: the filled-in timestamps are based on the mean/median dt, right? so if you have `[0, 1, 2, 3, 5]` in the first wafer, the algorithm would think the missing sample should be `4` when going with the median (or `4.25` when based on the mean). But it may turn out that all the other wafers actually had `[0, 1, 2, 3, 4.1, 5]`. The filled-in time value isn’t exactly what it should be. (This is a contrived example, but the idea is that your mean/median is only approximating where the true timestamps are)

But we’re going to be flagging those filled-in samples anyway, so maybe this doesn’t matter in the end.

File: sotodlib/io/imprinter.py
Comment (jlashner) [reply sotodlib/io/imprinter.py:764]: Sure, median would probably be more correct, I will update this. I hope large gaps like your describing would trigger some sort of warning somewhere.... Maybe we should do some sort of validation that tests that all arrays are indeed co-sampled before binding. That is not the purpose of this function but it might be a natural place to put it (like you mention above) since we're already extracting one set of timestamps.

For this scenario, I think we first have to figure out how we should actually handle this sort of event. If `dts/dt < 0.5` we may want to drop the sample. Like if we have `[..., 1, 2, 2.1, 3, 4]`, I think the resulting timestamps should still be [..., 1, 2, 3, 4]`

Comment (algadoreo) [reply sotodlib/io/imprinter.py:764]: Yeah those are definitely things to do down the line—no need to hold up this PR for them, I think. Nonetheless, the negative samples does mean I have to add a check in Bookbinder.

File: sotodlib/io/imprinter.py
Comment (jlashner) [reply sotodlib/io/imprinter.py:703]: Right... I'm not exactly sure what the correct behavior would be in this case.... maybe @mhasself can comment, I'm sure this has been thought through for other projects.

As far as I'm aware, these issues where the timing sample is off by a large amount and it's not due to a dropped sample are exceedingly rare / not possible, so I'm not sure how much we want to concern ourselves with those edge cases. If we are concerned... I think the right thing to do is to first decide how we want to be handling those cases, write tests, and then afterwards modify the function so it handles them properly.

File: sotodlib/io/imprinter.py
Comment (guanyilun) [inline sotodlib/io/imprinter.py:715]: I'm not sure we want to use book.start and book.stop to inform the min and max here, with the same reasoning as in the HWP PR that these timestamps are not accurate. You should only trust the timestamps from frames, so just point t0 and t1 to _t0 and _t1 in the first wafer would suffice.

**Review by guanyilun — COMMENTED** [review]
Sorry I don't really understand this `get_book_times` function. For example, how is returning a gap-filled timestamp array more helpful for bookbinder versus just passing the start and stop time and the sampling rate?

**Comment by jlashner** [comment]
We talked a bit about this on the call on Thursday, but in my opinion having a full array of all the timestamps that should be present in the book will greatly simplify the logic in the bookbinder. Currently, [the bookbinder is trying to detect and fill gaps](https://github.com/simonsobs/sotodlib/blob/c8bd0c41797dd73019fe86602340bfc0c6cd9ce9/sotodlib/io/bookbinder.py#L702-L743) after reading in each smurf frame individually, which is very difficult to get right since you have much less information. As Jason brought up, this is especially hard at the start/end of files and frames, and the current implementation may have [issues](https://github.com/simonsobs/sotodlib/issues/389).

I think it is more natural to have an initial step that first loads in the entire set of timestamps for an observation (assuming the runtime isn't too bad), and finding and filling gaps all at once using the function PR'ed here. Doing this in the imprinter, which has access to files from all UFMs, and passing the array of timestamps to the bookbinder will essentially eliminate the possibility of sampling errors like Matthew pointed out. This would be also be a natural place to add validation to check that the observations are actually co-sampled if we want to. 

Additionally, with the array of timestamps and the HK data, we have all we need to figure out the frame boundaries and the samples per frame for the entire book before we load in the smurf data. This could simplify the logic of the bookbinder, so it just needs to:
- allocate the array for a single output frame
- read in smurf data and put each sample into its closest slot
- interpolate in the slots that didn't get filled

Of course, you and Jason have thought about the bookbinder a lot more than I have, so you definitely don't have to use this tool if you think you have an easier solution to #389 that you'd rather implement.

**Comment by guanyilun** [comment]
Thanks for the clarification! I definitely think this function will be helpful for bookbinder for the reasons you listed and agree that imprinter is the best place for it. My comment earlier was more of an interface question, that is, I’m trying to understand the benefits of passing the entire timestamp array (ts) versus just ts[0], ts[-1] and np.median(np.diff(ts)) with which we can approximately reconstruct the entire array whenever needed. Naively I would expect them to be equivalent because we only care about the missing samples and they are gap-filled in your timestamp arrays based sampling rate anyway, but I have not thought about it deeply, maybe the timing system is not stable enough for this to be a robust method.

**Comment by jlashner** [comment]
Right, the issue is I don't think it's quite that simple to reconstruct `ts` from just `ts[0]`, `ts[-1]`, and the median sampling rate... 

If I compare the gap-filled timestamps I compute here `ts` with a reconstruction using 
```ts2 = np.arange(ts[0], ts[-1], np.median(np.diff(ts))```
you can see that the error from the median is blown up when you try to extrapolate to multi-hour long timestreams as is shown here:
image

There may be ways around this if you are more clever about how you're doing the reconstruction... for instance, computing the total number of samples in the gap-filled array and using `np.linspace` may work, but I think it is safer and easier to just pass in the full array of times, in which case you are sure that for the samples that are present, the timestamps are exactly the same as they are in the smurf files.

**Comment by guanyilun** [comment]
That’s convincing, thanks for the plot!

File: sotodlib/io/imprinter.py
Comment (kmharrington) [reply sotodlib/io/imprinter.py:692]: I don't think I've seen the trimming amount reported anywhere yet.

File: sotodlib/io/imprinter.py
Comment (kmharrington) [reply sotodlib/io/imprinter.py:725]: I approve of make this a general utility function and then only having one of them.

File: sotodlib/io/imprinter.py
Comment (mhasself) [reply sotodlib/io/imprinter.py:703]: The fill isn't based on median, actually ... it's a linterp, which will interpolate exactly between 3 and 5.  So you'll get 4.

However -- which of these new functions are designed to work on data where: (A) the timing system is in operation, and we can probably make multi-wafer books vs. (B) the timing system is not in operation.

The (A) case will be very stable... interpolating will bring only tiny errors, and I think that's fine.  Though you could pretty easily check+patch with the other vectors of timestreams, and this would be a great place to do it.

The (B) case can have very noisy timestamps... but of course in that case you don't need to worry about consistency between multiple wafers because you'll only have one wafer.

I don't think you need to have the "same code" run in cases (A) and (B).  It may be better to have separate, specialized functions for checking / filling / preparing the timestamp vectors in the two cases.

File: sotodlib/io/imprinter.py
Comment (jlashner) [reply sotodlib/io/imprinter.py:715]: I kind of thought this would be a good chance to enforce this, so as it is now the data will be trimmed so that all samples are within `book.start` and `book.stop`. I'm also fine with removing this though, but then we will always need to keep in mind that the `start` and `stop` are not hard limits.

Comment (guanyilun) [reply sotodlib/io/imprinter.py:715]: How about replacing book.start and book.end with the correct timestamps from this function?

Comment (jlashner) [reply sotodlib/io/imprinter.py:715]: I'll just leave it for now and not cut based on the book `start` and `stop`. I do think we should think about using the `get_start_and_end` function to set book.start and book.stop when the book is created though.

File: sotodlib/io/imprinter.py
Comment (jlashner) [reply sotodlib/io/imprinter.py:692]: Ok cool, I think I'm going to leave it for now and open an issue, since I think this should be done by the book-binder, which creates the M_index file

File: sotodlib/io/imprinter.py
Comment (jlashner) [reply sotodlib/io/imprinter.py:725]: Ok ya so I looked into this, and I think slight differences in the functions (i.e. Jason's function casts times to a G3VectorTime before returning) makes it difficult to use the exact same function for both purposes. I think the conversion from counters to timestamps however is a general function with a lot of hardcoded values that should probably be a shared util function. For now I separated this out into a new function `counters_to_timestamps` in the bookbinder, and use that function to construct the timestamps here.

**Comment by jlashner** [comment]
Hi all, thanks for the feedback! I added a commit that I think addresses the biggest points, so I think it's ready to be merged if someone approves it. I think things like tests for handling edge-cases should be added in a separate PR.

**Review by guanyilun — APPROVED** [review]

**Comment by algadoreo** [comment]
Wait—in `fill_time_gaps`, L773, weren’t we changing to the median dt instead of the mean?

**Comment by jlashner** [comment]
Oops sorry, merged this right before I saw your comment Jason, let me get that

**Pull request merged**
